packages/ember-data/lib/system/model/internal-model.js

```diff
@@ -0,0 +1,696 @@
+    attributes and relationships to be preloaded. These are meant to behave as if they
+    came back from the server, except the user obtained them out of band and is informing
+    the store of their existence. The most common use case is for supporting client side
+    nested URLs, such as `/posts/1/comments/2` so the user can do
+    `store.find('comment', 2, {post:1})` without having to fetch the post.
+
+    Preloaded data can be attributes and relationships passed in either as IDs or as actual
+    models.
+
+    @method _preloadData
+    @private
+    @param {Object} preload
+  */
+  _preloadData: function(preload) {
+    var record = this;
+    //TODO(Igor) consider the polymorphic case
+    forEach.call(Ember.keys(preload), function(key) {
+      var preloadValue = get(preload, key);
+      var relationshipMeta = record.type.metaForProperty(key);
+      if (relationshipMeta.isRelationship) {
+        record._preloadRelationship(key, preloadValue);
+      } else {
+        get(record, '_data')[key] = preloadValue;
+      }
+    });
+  },
+
+  _preloadRelationship: function(key, preloadValue) {
+    var relationshipMeta = this.type.metaForProperty(key);
+    var type = relationshipMeta.type;
+    if (relationshipMeta.kind === 'hasMany') {
+      this._preloadHasMany(key, preloadValue, type);
+    } else {
+      this._preloadBelongsTo(key, preloadValue, type);
+    }
+  },
+
+  _preloadHasMany: function(key, preloadValue, type) {
+    Ember.assert("You need to pass in an array to set a hasMany property on a record", Ember.isArray(preloadValue));
+    var record = this;
+
+    var recordsToSet = map.call(preloadValue, function(recordToPush) {
+      return record._convertStringOrNumberIntoInternalModel(recordToPush, type);
+    });
+    //We use the pathway of setting the hasMany as if it came from the adapter
+    //because the user told us that they know this relationships exists already
+    this._relationships[key].updateRecordsFromAdapter(recordsToSet);
+  },
+
+  _preloadBelongsTo: function(key, preloadValue, type) {
+    var recordToSet = this._convertStringOrNumberIntoInternalModel(preloadValue, type);
+
+    //We use the pathway of setting the hasMany as if it came from the adapter
+    //because the user told us that they know this relationships exists already
+    this._relationships[key].setRecord(recordToSet);
+  },
+
+  _convertStringOrNumberIntoInternalModel: function(value, type) {
+    if (Ember.typeOf(value) === 'string' || Ember.typeOf(value) === 'number') {
+      return this.store._internalModelForId(type, value);
+    }
+    if (value._internalModel) {
+      return value._internalModel;
+    }
+    return value;
+  },
+
+
+  /**
+    @method updateRecordArrays
+    @private
+  */
+  updateRecordArrays: function() {
+    this._updatingRecordArraysLater = false;
+    this.store.dataWasUpdated(this.type, this);
+  },
+
+  setId: function(id) {
+    this.id = id;
```

Comment:
seems like it might be risky to keep this state in two places. Although that isn't true, if ID is now immutable.